File: analysis/compute_mean_time.m
function [ mean_time ] = compute_mean_time( S, dT )
%COMPUTE_MEAN_TIME Calculates the "mean time" for a signal
%   The mean time is defined as the center of mass of the time curve.
% INPUTS
%	S - signal dynamics, time must be in the last dimension [voxels and/or # of metabolites, # of time points]
%   dT - spacing of time points
% OUTPUTS
%   mean_time - mean time computed for all voxels and/or metabolites 

Ns = size(S);
tdim = length(Ns);
Nt = Ns(tdim);

if length(Ns) > 1
    tmat = repmat(reshape( (0:Nt-1)*dT, [ones(1,length(Ns)-1),Nt]), [Ns(1:end-1),1]);
else
    tmat = (0:Nt-1)*dT;
end

mean_time = sum( S .* tmat,tdim) ./ sum(S,tdim);

end

File: analysis/python/compute_mean_time.py
import numpy as np

def compute_mean_time(S, dT):
    """
    Computes the mean time for a signal, defined as the center of mass of the time curve.
    
    Parameters:
    S : numpy array
        Signal dynamics, with time in the last dimension [voxels and/or # of metabolites, # of time points].
    dT : float
        Spacing of time points.
    
    Returns:
    numpy array
        Mean time computed for all voxels and/or metabolites.
    """
    S_shape = S.shape
    Nt = S_shape[-1]  # Number of time points
    
    t_values = np.arange(Nt) * dT
    
    if len(S_shape) > 1:
        tmat = np.reshape(t_values, [*(1,) * (len(S_shape) - 1), Nt])
        tmat = np.broadcast_to(tmat, S_shape)  # Match dimensions of S
    else:
        tmat = t_values
    
    mean_time = np.sum(S * tmat, axis=-1) / np.sum(S, axis=-1)
    
    return mean_time
